### PR TITLE
fix: do not unnecessarily wait on additional snapshot in server request handlers

### PR DESCRIPTION
**Pull request by Kha**
I'm afraid we do not have any sufficient infrastructure for writing a test for this

### CODE REVIEW

File: src/Lean/Server/AsyncList.lean [hunk 1]
```diff
@@ -65,19 +65,23 @@ partial def getAll : AsyncList ε α → List α × Option ε
...
-partial def waitAll (p : α → Bool := fun _ => true) : AsyncList ε α → Task (List α × Option ε)
+/-- Spawns a `Task` returning the prefix of elements up to (including) the first element for which `p` is true. -/
+partial def waitUntil (p : α → Bool) : AsyncList ε α → Task (List α × Option ε)
   | cons hd tl =>
-    if p hd then
-      (tl.waitAll p).map fun ⟨l, e?⟩ => ⟨hd :: l, e?⟩
+    if !p hd then
+      (tl.waitUntil p).map fun ⟨l, e?⟩ => ⟨hd :: l, e?⟩
     else
       .pure ⟨[hd], none⟩
```
Comment (Kha) [inline src/Lean/Server/AsyncList.lean:74]: Ironically enough it already returned this additional element, no change necessary

File: src/Lean/Server/AsyncList.lean [hunk 2]
```diff
@@ -65,19 +65,23 @@ partial def getAll : AsyncList ε α → List α × Option ε
...
   | nil => .pure ⟨[], none⟩
   | delayed tl =>
     tl.bind fun
-      | .ok tl   => tl.waitAll p
+      | .ok tl   => tl.waitUntil p
       | .error e => .pure ⟨[], some e⟩
 
+/-- Spawns a `Task` waiting on all elements. -/
+partial def waitAll : AsyncList ε α → Task (List α × Option ε) :=
```
Comment (Vtec234) [inline src/Lean/Server/AsyncList.lean:82]:
```suggestion
def waitAll : AsyncList ε α → Task (List α × Option ε) :=
```

File: src/Lean/Server/AsyncList.lean [hunk 3]
```diff
@@ -65,19 +65,23 @@ partial def getAll : AsyncList ε α → List α × Option ε
     | Except.ok tl => tl.getAll
     | Except.error e => ⟨[], some e⟩
 
-/-- Spawns a `Task` waiting on the prefix of elements for which `p` is true. -/
-partial def waitAll (p : α → Bool := fun _ => true) : AsyncList ε α → Task (List α × Option ε)
+/-- Spawns a `Task` returning the prefix of elements up to (including) the first element for which `p` is true. -/
```
Comment (Vtec234) [inline src/Lean/Server/AsyncList.lean:68]:
```suggestion
/-- Spawns a `Task` returning the prefix of elements up to (including) the first element for which `p` is true.
When `p` is not true of any element, it returns the entire list. -/
```

**Review by Vtec234 — APPROVED** [review]
Ah, a better API!

**Pull request merged**